pkg/operation/botanist/addons.go

```diff
@@ -172,7 +176,123 @@ func (b *Botanist) DeployManagedResources(ctx context.Context) error {
+	}
+	cloudConfigManagedResource := manager.
+		NewManagedResource(b.K8sSeedClient.Client()).
+		WithNamespacedName(b.Shoot.SeedNamespace, managedResourceName).
+		WithLabels(labels).
+		WithInjectedLabels(injectedLabels).
+		KeepObjects(false)
+
+	// reconcile secrets and reference them to the ManagedResource
+	fns := make([]flow.TaskFn, 0, len(cloudConfigCharts))
+	for name, renderChartFunc := range cloudConfigCharts {
+		renderedChart, err := renderChartFunc()
+		if err != nil {
+			return fmt.Errorf("error rendering %q chart: %+v", name, err)
+		}
+		data := renderedChart.AsSecretData()
+		secretName := "managedresource-" + name
+		fns = append(fns, func(ctx context.Context) error {
+			return manager.
+				NewSecret(b.K8sSeedClient.Client()).
+				WithNamespacedName(b.Shoot.SeedNamespace, secretName).
+				WithKeyValues(data).
+				WithLabels(secretLabels).
+				Reconcile(ctx)
+		})
+		cloudConfigManagedResource.WithSecretRef(secretName)
+		wantedSecretNames.Insert(secretName)
+	}
+	if err := flow.Parallel(fns...)(ctx); err != nil {
+		return err
+	}
+	if err := cloudConfigManagedResource.Reconcile(ctx); err != nil {
+		return err
+	}
+	if err := b.deleteStaleSecretsMatchLabel(ctx, secretLabels, wantedSecretNames); err != nil {
+		return err
+	}
+
+	// TODO: remove this in the future
+	return b.deleteLegacyCloudConfigSecret(ctx)
+}
+
+// deleteStaleSecretsMatchLabel deletes the stale secrets that match the labels but are not used anymore
+func (b *Botanist) deleteStaleSecretsMatchLabel(ctx context.Context, labels map[string]string, wantedSecretNames sets.String) error {
+	c := b.K8sSeedClient.Client()
+
+	secretList := &corev1.SecretList{}
+	if err := c.List(ctx, secretList, client.InNamespace(b.Shoot.SeedNamespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	fns := make([]flow.TaskFn, 0, meta.LenList(secretList))
+	for _, secret := range secretList.Items {
+		if !wantedSecretNames.Has(secret.Name) {
+			toDelete := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secret.Name,
+					Namespace: secret.Namespace,
+				},
+			}
+			fns = append(fns, func(ctx context.Context) error {
+				if err := common.ConfirmDeletion(ctx, c, toDelete); err != nil {
+					return err
+				}
+				return client.IgnoreNotFound(c.Delete(ctx, toDelete, kubernetes.DefaultDeleteOptions...))
+			})
+		}
+	}
+	return flow.Parallel(fns...)(ctx)
+}
+
+// deleteLegacyCloudConfigSecret deletes the legacy secret created prior to v1.3.0, which is not labeled with the referenced ManagedResource
+func (b *Botanist) deleteLegacyCloudConfigSecret(ctx context.Context) error {
+	c := b.K8sSeedClient.Client()
+
+	legacySecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "managedresource-shoot-cloud-config-execution",
+			Namespace: b.Shoot.SeedNamespace,
+		},
+	}
+	if err := common.ConfirmDeletion(ctx, b.K8sSeedClient.Client(), legacySecret); err != nil {
```

Comment:
This is not needed for the secret  deletion

Reply:
Addressed in https://github.com/gardener/gardener/pull/2210/commits/b850b90295d18f227fa810fa08b1e67e5661d075